### ISSUE TITLE
Making puppet-nginx work with puppet-letsencrypt in webroot mode

### ISSUE THREAD

**Issue by kwisatz** [voxpupuli/puppet-nginx#1123]
I wanted to see if someone else already solved this problem or, if not, discuss it so that we may find a viable and functional solution.

We're in a HA setup were using any lenc plugin other than `webroot` for renewal and certificate creation is not an option.

However, getting to a working configuration with puppet is now a two-step process for us, which kind of defeats the purpose of using puppet in the first place:

1. Define the server in nginx to run on `port 80` _-> run puppet_
1. Define the letsencrypt configuration to have a certificate be generated and enable `ssl` on nginx _-> run puppet again_

This is necessary since we have a circular dependency here:

- let's encrypt with `webroot` depends on the .well-known location having been defined and being available
- nginx ssl depends on the certificate and key files already existing

To the best of my knowledge, both can't currently be satisfied in a single puppet agent run.

If I understand the anchor resource correctly – and I'm not sure I do – a possible solution would be to chain the execution flow like this:

1. realize nginx server on `port 80` including `webroot location`
1. realize letsencrypt certificate generation
1. realize nginx server `443 ssl;` configuration

I'm not entirely sure yet how one would go about implementing this or if this makes any sense in the first place, but I'm very keen on seeing if anyone has had the same problem and has either solved it (differently) or is interested in discussing and coming up with a solution.

**Comment by ekohl** [comment]
You're right: this is not something puppet can easily do. You can think of some workarounds, for example create a custom fact ```enable_ssl``` which looks at if a certificate has been requested. Then you only create the SSL-enabled vhosts if that fact is true. Then you need 2 runs but at least it's automated.

A more complicated solution would be to chain the resources in such a way that you can start the server with an exec. It is possible that the following works but you will have double nginx reloads/restarts when a non-ssl vhost changes.

```puppet
exec { 'ensure nginx':
  command     => '/bin/systemctl reload-or-restart nginx':
  refreshonly => true,
  require     => Class['nginx::config'],
}

Nginx::Resource::Server <| ssl == false |> ~> Exec['ensure nginx'] -> Class['letsencrypt'] -> Nginx::Resource::Server <| ssl == true |>
```

**Comment by kwisatz** [comment]
Hi, @ekohl, thanks for your input! 

Please CMIIW:

As to your **first suggestion**: With a set-up that relies heavily – if not exclusively – on hiera and auto-lookup like ours, this would either require support for this new fact inside the module's manifests, or it would mean stamping out custom modules that more or less re-produce the module's `resource/server` class augmented by checks for the custom fact.

Again, making heavy use of hiera, the **second suggestion** requires us to write additional code in order to create virtual resources from the hiera data where previously, an `include nginx` would have been sufficient. Here I'd probably replace the module's `init`class with a custom one that creates virtual resources instead of realizing them on the spot.

In both cases, we'd be writing code that would have to be maintained alongside the upstream module, something that everyone would probably prefer to avoid.

**Comment by ekohl** [comment]
> As to your first suggestion: With a set-up that relies heavily – if not exclusively – on hiera and auto-lookup like ours, this would either require support for this new fact inside the module's manifests, or it would mean stamping out custom modules that more or less re-produce the module's resource/server class augmented by checks for the custom fact.

I agree that this is a bad solution. I also didn't think about a situation where you add a new vhost during the lifetime and rely on certificates to exist.

> Again, making heavy use of hiera, the second suggestion requires us to write additional code in order to create virtual resources from the hiera data where previously, an include nginx would have been sufficient. Here I'd probably replace the module's initclass with a custom one that creates virtual resources instead of realizing them on the spot.

What I suggested can be written as a profile class somewhere and would work with other includes. The collectors I wrote are all local, no virtual resources or exported resources. Now I see it should most likely be a collector for ```Letsencrypt::Certonly```.

> In both cases, we'd be writing code that would have to be maintained alongside the upstream module, something that everyone would probably prefer to avoid.

That would be some coupling with another module and I don't know how other maintainers feel about this. Personally I prefer the pattern where the coupling is in the profiles, not plain modules. This is more business logic. On the other hand it would be great if this was an example somewhere, ideally in acceptance tests.

**Comment by wyardley** [comment]
Maybe this is one of those rare times to use stages (in your site specific code, not in the module)?  Or setup the domain with a dummy cert (with both SSL and no SSL enabled) until the process is completed?

I haven't used Letsencrypt much, but I noticed that they also support using DNS records for validation, maybe that would be easier to setup (when new domains / subdomains are provisioned), so that you don't have to setup a non-SSL server block just to handle the process.

It would be cool if the module itself could have better support for this, but from a quick description of how the process works, it seems like it might be difficult.

**Comment by ekohl** [comment]
> I haven't used Letsencrypt much, but I noticed that they also support using DNS records for validation, maybe that would be easier to setup (when new domains / subdomains are provisioned), so that you don't have to setup a non-SSL server block just to handle the process.

That's a great feature but in many cases it's hard to use because the DNS-provider lacks a proper API.

**Comment by ekohl** [comment]
I do agree we could include an optional class if it works.

**Comment by law** [comment]
Howdy!  I have similar issues, and IMHO it stems from nginx's cowardly refusal to start when an SSL certificate defined in its config does not exist yet.  I handle this in the following way: 

1. include the LE puppet module (puts the LE client on the host)
2. use an Exec to one-shot the LE client in 'standalone' mode to get the initial cert
3. define the letsencrypt cert parameters with the 'webroot' plugin
4. include the nginx module, define vhosts and locations as needed

From a code perspective, it looks like this: 
```
class role::webserver (
  $nginx_vhosts,
  $nginx_location,
  $letsencrypt_certs
) {
  class { 'letsencrypt': }
  class { 'nginx': }

  exec { 'first-run LE to get certs': # This only runs once, per the 'creates' parameter
    command => "/usr/bin/letsencrypt --text --agree-tos certonly -a standalone -d myhost.example.com",
    creates => "/etc/letsencrypt/live/myhost.example.com/fullchain.pem",
  }

  create_resources('letsencrypt::certonly', $letsencrypt_certs)
  create_resources('nginx::resource::server', $nginx_vhosts)
  create_resources('nginx::resource::location', $nginx_location)
}
```

Hieradata: 
```
---
role::webserver::letsencrypt_certs:
  'mycerts':
    domains: ["myhost.example.com"]
    manage_cron: true
    plugin: 'webroot'
    webroot_paths: ['/var/www/letsencrypt']
    manage_cron: true
    cron_success_command: '/bin/systemctl restart nginx'

role::webserver::nginx_vhosts:
  'myhost.example.com':
    www_root: '/var/www/example.com'
    server_name: ['myhost.example.com']
    access_log: '/var/log/nginx/example.access.log'
    error_log: '/var/log/nginx/example.error.log'
    ssl_redirect: true
    ssl: true
    ssl_cert: '/etc/letsencrypt/live/myhost.example.com/fullchain.pem'
    ssl_key: '/etc/letsencrypt/live/myhost.example.com/privkey.pem'

role::webserver::nginx_location: # this is needed for renewing the LE certs
  'letsencrypt-well-known':
    location: '/.well-known'
    www_root: '/var/www/letsencrypt'
    try_files: ['$uri $uri/ =404']
    ssl_only: false   # NB: LE renewal requests that get 301'd to https will fail
    server: "myhost.example.com"
```

This Works For Me(tm), but I had to make a slight modification to the templates/server/server_header.erb template to get LE renewing cleanly.  See https://github.com/voxpupuli/puppet-nginx/pull/1127 if you want to add it yourself, or wait to see how the discussion shakes out and it may make it into the next release.  

HTH, HAND, etc.

**Comment by ekohl** [comment]
@law that's a good idea but shouldn't you add a before on the exec so you're sure it executes before nginx?

**Comment by law** [comment]
You are absolutely right, a `before` is more than appropriate here rather than relying on manifest-order.  Pull-request received, reviewed, and accepted with thanks!  :-)

**Comment by khaefeli** [comment]
I think the `ssl_redirect = true` in the hieradata creates a http to https redirect on the server{} level.
this causes that all requests are redirected to https. Remove this line and it should work
